// @nrwl/angular/src/schematics/upgrade-module/upgrade-module.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const testing_1 = require("@nrwl/workspace/testing");
const testing_2 = require("../../utils/testing");
const workspace_1 = require("@nrwl/workspace");
describe('upgrade-module', () => {
    let appTree;
    beforeEach(() => {
        appTree = schematics_1.Tree.empty();
        appTree = testing_1.createEmptyWorkspace(appTree);
        appTree = testing_2.createApp(appTree, 'myapp');
    });
    it('should update the bootstrap logic', () => __awaiter(void 0, void 0, void 0, function* () {
        const tree = yield testing_2.runSchematic('upgrade-module', {
            name: 'legacy',
            project: 'myapp'
        }, appTree);
        const appModule = testing_1.getFileContent(tree, '/apps/myapp/src/app/app.module.ts');
        expect(appModule).toContain(`this.upgrade.bootstrap(document.body, ['downgraded', 'legacy'])`);
        expect(appModule).not.toContain(`bootstrap:`);
        const legacySetup = testing_1.getFileContent(tree, '/apps/myapp/src/legacy-setup.ts');
        expect(legacySetup).toContain(`import 'legacy';`);
        expect(tree.exists('/apps/myapp/src/hybrid.spec.ts')).toBeTruthy();
    }));
    it('should update package.json by default', () => __awaiter(void 0, void 0, void 0, function* () {
        appTree.overwrite(`/package.json`, JSON.stringify({
            dependencies: {
                '@angular/core': '4.4.4'
            }
        }));
        const tree = yield testing_2.runSchematic('upgrade-module', {
            name: 'legacy',
            project: 'myapp'
        }, appTree);
        const packageJson = workspace_1.readJsonInTree(tree, '/package.json');
        expect(packageJson.dependencies['@angular/upgrade']).toEqual('4.4.4');
        expect(packageJson.dependencies['angular']).toBeDefined();
    }));
    it('should not package.json when --skipPackageJson=true', () => __awaiter(void 0, void 0, void 0, function* () {
        appTree.overwrite(`/package.json`, JSON.stringify({
            dependencies: {
                '@angular/core': '4.4.4'
            }
        }));
        const tree = yield testing_2.runSchematic('upgrade-module', {
            name: 'legacy',
            skipPackageJson: true,
            project: 'myapp'
        }, appTree);
        const packageJson = workspace_1.readJsonInTree(tree, '/package.json');
        expect(packageJson.dependencies['@angular/upgrade']).not.toBeDefined();
    }));
    it('should add router configuration when --router=true', () => __awaiter(void 0, void 0, void 0, function* () {
        const tree = yield testing_2.runSchematic('upgrade-module', {
            name: 'legacy',
            router: true,
            project: 'myapp'
        }, appTree);
        const legacySetup = testing_1.getFileContent(tree, '/apps/myapp/src/legacy-setup.ts');
        expect(legacySetup).toContain(`setUpLocationSync`);
    }));
    it('should support custom angularJsImport', () => __awaiter(void 0, void 0, void 0, function* () {
        const tree = yield testing_2.runSchematic('upgrade-module', {
            name: 'legacy',
            angularJsImport: 'legacy-app',
            project: 'myapp'
        }, appTree);
        const legacySetup = testing_1.getFileContent(tree, '/apps/myapp/src/legacy-setup.ts');
        expect(legacySetup).toContain(`import 'legacy-app';`);
        expect(legacySetup).not.toContain(`import 'legacy';`);
    }));
});
